### PR TITLE
router: adding a knob to configure a cap on buffering for shadowing/retries

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: Medium (refactors to existing router logic)
Testing: new unit tests
Docs Changes: n/a
Release Notes: added

### CODE REVIEW

File: source/common/router/router.cc [hunk 1]
```diff
@@ -662,7 +664,9 @@ void Filter::setDecoderFilterCallbacks(Http::StreamDecoderFilterCallbacks& callb
   // As the decoder filter only pushes back via watermarks once data has reached
   // it, it can latch the current buffer limit and does not need to update the
   // limit if another filter increases it.
-  buffer_limit_ = callbacks_->decoderBufferLimit();
+  if (callbacks_->decoderBufferLimit() != 0) {
+    retry_shadow_buffer_limit_ = callbacks_->decoderBufferLimit();
```
Comment (snowp) [inline source/common/router/router.cc]: its a little bit hard to follow that this limit is set to the buffer limit in the callbacks, then potentially updated in `decodeHeaders` to account for the retry/shadow limit. Maybe a comment here would help?

File: api/envoy/api/v2/route/route.proto [hunk 2]
```diff
@@ -230,6 +230,11 @@ message Route {
   // Presence of the object defines whether the connection manager's tracing configuration
   // is overridden by this route specific instance.
   Tracing tracing = 15;
+
+  // The maximum bytes which will be buffered for retries and shadowing.
+  // The bytes actually buffered will be the minimum value of this and the
+  // listener per_connection_buffer_limit_bytes.
+  google.protobuf.UInt32Value per_request_buffer_limit_bytes = 16;
```
Comment (htuch) [inline api/envoy/api/v2/route/route.proto]: Would it make sense to have this at multiple levels, e.g. `RouteConfiguration` and/or `VirtualHost`. That way, a default can be set that is "safe", without having to do it for every route, and when you have routes with special requirements, e.g. large payloads or streams, we can also override at the route level?

**Comment by mattklein123** [comment]
Oops looks like this needs master merge.

/wait

File: test/common/http/async_client_impl_test.cc [hunk 3]
```diff
@@ -1183,46 +1184,57 @@ TEST_F(AsyncClientImplTest, DumpState) {
...
-  EXPECT_EQ(nullptr, route_impl.routeEntry()->metadataMatchCriteria());
-  EXPECT_TRUE(route_impl.routeEntry()->rateLimitPolicy().empty());
-  EXPECT_TRUE(route_impl.routeEntry()->rateLimitPolicy().getApplicableRateLimit(0).empty());
-  EXPECT_EQ(absl::nullopt, route_impl.routeEntry()->idleTimeout());
-  EXPECT_EQ(absl::nullopt, route_impl.routeEntry()->grpcTimeoutOffset());
-  EXPECT_TRUE(route_impl.routeEntry()->opaqueConfig().empty());
-  EXPECT_TRUE(route_impl.routeEntry()->includeVirtualHostRateLimits());
-  EXPECT_TRUE(route_impl.routeEntry()->metadata().filter_metadata().empty());
+TEST_F(AsyncClientImplUnitTest, RouteImplInitTest) {
+
```
Comment (mattklein123) [inline test/common/http/async_client_impl_test.cc]: nit: extra line

**Comment by alyssawilk** [comment]
ugh, still conflicting, and fix_format is broken locally
